### DISCUSSION TITLE
Cuda support?

### DISCUSSION THREAD

**Discussion by kristopher-wood** [online-ml/river#663] [Q&A]
Does River have any Cuda support? It would be nice if I could fully utilize my GPU.

**Comment by chrsunwil** [comment]
Hi, I would love to bump this question. I assume that there would still be some speed up for these models (although I know that the largest speed-ups are general for neural networks).

**Reply by MaxHalford** [reply]
Hello there. We don't offer CUDA support because River models are meant to learn with single instances. Using GPUs would bring too much overhead. But I'd be happy to be proved wrong.

**Reply by chrsunwil** [reply]
Thanks, that makes sense to me and, admittedly, I'm not very familiar with the streaming implementations of many of these models, but would there not be any speed gains for larger ensembles? I guess I should think about it more, but I was thinking there would be gains for adaboost. I guess not though because there wouldn't be any matrix multiplication (because single samples). 

My adaboost is taking relatively long to train on some samples (obviously depends on the error, etc.), any advice for speeding up besides reducing the number of trees/complexity of data?

**Reply by MaxHalford** [reply]
There are likely gains to be made by adding some parallelism in there. To confirm, do you notice if your CPU usage is close to a 100% or not?

**Reply by chrsunwil** [reply]
My CPU is only at about 23%, but I wasn't sure how much parallelism was built into the implementation? Also, to be fair, I'm training two models so I can parallelize my surrounding code a bit.

**Reply by MaxHalford** [reply]
There is no parallelism built in as of now :)

**Reply by chrsunwil** [reply]
Even for bagging ensembles? (Those are embarrassingly parallel, right?). Because my CPU is only 23%, that indicates there could be gains from parallelization, right? 

As a side note, thanks for maintaining this package! I just discovered it a handful of weeks ago and it's really wonderful.

**Reply by MaxHalford** [reply]
> Even for bagging ensembles? (Those are embarrassingly parallel, right?). Because my CPU is only 23%, that indicates there could be gains from parallelization, right?

There should be, in theory. But then in reality you have a communication cost: each observation needs to be sent to a worker for processing. Sometimes that results in too much overhead. There's no silver bullet, especially for small processing loads, as is the case with online machine learning!

> As a side note, thanks for maintaining this package! I just discovered it a handful of weeks ago and it's really wonderful.

Cheers, we're doing what we can :)
